ThirdPartyAdapters/chartboost/chartboost/src/main/java/com/google/ads/mediation/chartboost/ChartboostAdapter.java

```diff
@@ -186,6 +186,101 @@ public void onPause() {
+                                MediationBannerListener mediationBannerListener,
+                                Bundle serverParameters,
+                                AdSize adSize,
+                                MediationAdRequest mediationAdRequest,
+                                Bundle networkExtras) {
+        mMediationBannerListener = mediationBannerListener;
+        mChartboostParams = ChartboostAdapterUtils.createChartboostParams(serverParameters, networkExtras);
+        if (!ChartboostAdapterUtils.isValidChartboostParams(mChartboostParams)) {
+            // Invalid server parameters, send ad failed to load event.
+            if (mMediationBannerListener != null) {
+                mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this,
+                        AdRequest.ERROR_CODE_INVALID_REQUEST);
+            }
+            return;
+        }
+        ChartboostSingleton.startChartboostBanner(context, mChartboostBannerDelegate, createChartboostBannerListener());
+    }
+
+    @Override
+    public View getBannerView() {
+        WeakReference<ChartboostBanner> ref = ChartboostSingleton.getChartboostBanner(mChartboostParams.getLocation());
+        if(ref != null) {
+            return ref.get();
+        }
+        return null;
+    }
+
+    private ChartboostBannerListener createChartboostBannerListener() {
+        return new ChartboostBannerListener() {
+            @Override
+            public void onAdCached(ChartboostCacheEvent chartboostCacheEvent, ChartboostCacheError chartboostCacheError) {
+                if(mMediationBannerListener!=null) {
+                    if(chartboostCacheError == null) {
+                        mMediationBannerListener.onAdLoaded(ChartboostAdapter.this);
+                    } else {
+                        reportChartboostErrorToAdMob(chartboostCacheError.code);
+                    }
+                }
+            }
+
+            @Override
+            public void onAdShown(ChartboostShowEvent chartboostShowEvent, ChartboostShowError chartboostShowError) {
+                if(mMediationBannerListener!=null) {
+                    if (chartboostShowError == null) {
+                        mMediationBannerListener.onAdOpened(ChartboostAdapter.this);
+                    } else {
+                        reportChartboostErrorToAdMob(chartboostShowError.code);
+                    }
+                }
+            }
+
+            @Override
+            public void onAdClicked(ChartboostClickEvent chartboostClickEvent, ChartboostClickError chartboostClickError) {
+                if (mMediationBannerListener != null) {
+                    if (chartboostClickError == null) {
+                        mMediationBannerListener.onAdClicked(ChartboostAdapter.this);
+                        mMediationBannerListener.onAdLeftApplication(ChartboostAdapter.this);
+                    } else {
+                        reportChartboostErrorToAdMob(chartboostClickError.code);
```

Comment:
Are you sure you want to report an ad load failure error if click fails?
I think AdMob doesn't have a clickFailed method, so we could just do nothing in this case

Reply:
I just want to report our error so i think if our on click gets an error we should report it.

Reply:
I don't think this is the proper way though. Reporting a load failure may cause AdMob to close our ad our do something weird. If AdMob doesn't know about click failures (their public listener methods don't have anything like it) then our click failures will be lost anyway.
That's fine though, it's not critical for anyone. The only important thing is that the click is reported to AdMob when we report it to AdServer.

Reply:
Fixed